### ISSUE TITLE
Generate password complexity setting

### ISSUE THREAD

**Issue by zapper** [phpmyadmin/phpmyadmin#15491]
Hello,

I always encounter problem when mysql server is set to medium password level it requires special character in passwords when creating user.

When you use generate password function it doesn't have any of special chars so you always get an error from mysql for incorrect password.

Solution was simple: add special chars to function suggestPassword in functions.js file, no commas, no dots, no quotes, just these special chars !@#$%^&*()

How about making a setting in phpmyadmin to let people choose password complexity without having to modify js files? :)

**Comment by OlafvdSpek** [comment]
Can't MySQL be fixed to not require special chars per se? ;)

**Comment by zapper** [comment]
> Can't MySQL be fixed to not require special chars per se? ;)

Do you think that lowering security level for passwords is a good way to solve the problem? :)

**Comment by zapper** [comment]
> Who's talking about lowering security levels?
> 
> Each 'security level' can be reached without special characters, simply by increasing the password length.

Personally I don't see any reasons why it can't be added.

**Comment by Tithugues** [comment]
Just a note, but it might be interesting to generate a password regarding the MySQL configuration : https://dev.mysql.com/doc/refman/5.7/en/validate-password-options-variables.html#sysvar_validate_password_policy

**Comment by Tithugues** [comment]
> > Who's talking about lowering security levels?
> > Each 'security level' can be reached without special characters, simply by increasing the password length.
> 
> Personally I don't see any reasons why it can't be added.

In the code, there is one reason:

> restrict the password to just letters and numbers to avoid problems:
"editors and viewers regard the password as multiple words and
things like double click no longer work"

In my opinion, this is not a good reason…

**Comment by OlafvdSpek** [comment]
That doesn't add a setting for password complexity, it changes it unconditionally. :(

**Comment by Tithugues** [comment]
@OlafvdSpek , you're right. My first idea was to provide a quick fix to be able to generate a valid password (even if it may not always be valid). But here, at least, we **can** have a valid password.

In another version, I would like to generate a password regarding the MySQL configuration.

So I would suggest to keep this issue open if this is OK for you @williamdes @mauriciofauth .

Thank you.

**Comment by williamdes** [comment]
Okay
I did not clear the milestone since I hope someone will do the implementation before the release :)
Thank you for the quick fix @Tithugues

**Comment by OlafvdSpek** [comment]
Could the update be reverted until it's (fully) ready?

**Comment by Tithugues** [comment]
@OlafvdSpek, sorry, but why ?

Is there any issue with this patch ? Because there might be a better solution, does it mean that we shouldn't provide a temporary solution ?

Thank you for your feedback.

**Comment by OlafvdSpek** [comment]
See the reason for the current solution @ https://github.com/phpmyadmin/phpmyadmin/issues/15491#issuecomment-546742667

**Comment by rabeehrz** [comment]
Hi, I’m new here. I would love to contribute to phpMyAdmin and I think this is an issue I can work on. Can someone tell me what exactly I should implement?

P.S: I’m just starting open source contribution and I might be a bit slow, but I love to learn! So.

**Comment by Tithugues** [comment]
Hi,

I wanted to work on it, but didn't yet find time to do it…

Here is the way I expected to do it:
* Retrieve all MySQL password policy variables
* Share them with JS (as it's done for "Messages")
* Use these variables in the current JavaScript function to generate the new password
* Smile

**Comment by rabeehrz** [comment]
Sure, Will work on it. Thanks! :D

**Comment by Tithugues** [comment]
For information, here is the list of variables: https://dev.mysql.com/doc/refman/5.7/en/validate-password-options-variables.html

**Comment by rabeehrz** [comment]
Yeah, I saw that earlier. And I did devise a solution exactly the same as yours. But I didn't know how to share those variables with JS. Since somewhat similar is already implemented, I'll look the code and make something similar.
Thanks

**Comment by rabeehrz** [comment]
> * Share them with JS (as it's done for "Messages")

Can you show me the file where this is done, I'm having a hard time finding it. :sweat_smile: 
Thanks

**Comment by Tithugues** [comment]
I was thinking about this file: https://github.com/phpmyadmin/phpmyadmin/blob/master/js/messages.php

**Comment by rabeehrz** [comment]
Okay Thanks, I'll look into this! :smile:

**Comment by rabeehrz** [comment]
Can some one help me with this? I need a way to retrieve the MySQL password policies and share them with the function. How can I do it?

**Comment by williamdes** [comment]
Please refer to https://github.com/phpmyadmin/phpmyadmin/issues/16305 for more informations
